base/setup/lib/bootsup.c

```diff
@@ -90,544 +119,414 @@ typedef struct _EXT2_BOOTSECTOR
 
 #include <poppack.h>
 
-extern PPARTLIST PartitionList;
+/* End of BIG FIXME!! */
+
 
 /* FUNCTIONS ****************************************************************/
 
+static VOID
+TrimTrailingPathSeparators_UStr(
+    IN OUT PUNICODE_STRING UnicodeString)
+{
+    while (UnicodeString->Length >= sizeof(WCHAR) &&
+           UnicodeString->Buffer[UnicodeString->Length / sizeof(WCHAR) - 1] == OBJ_NAME_PATH_SEPARATOR)
+    {
+        UnicodeString->Length -= sizeof(WCHAR);
+    }
+}
+
 
-static
-VOID
-CreateCommonFreeLoaderSections(
-    PINICACHE IniCache)
+static VOID
+CreateFreeLoaderReactOSEntries(
+    IN PVOID BootStoreHandle,
+    IN PCWSTR ArcPath)
 {
-    PINICACHESECTION IniSection;
+    UCHAR xxBootEntry[FIELD_OFFSET(BOOT_STORE_ENTRY, OsOptions) + sizeof(NTOS_OPTIONS)];
+    PBOOT_STORE_ENTRY BootEntry = (PBOOT_STORE_ENTRY)&xxBootEntry;
+    PNTOS_OPTIONS Options = (PNTOS_OPTIONS)&BootEntry->OsOptions;
+    BOOT_STORE_OPTIONS BootOptions;
+
+    BootEntry->Version = FreeLdr;
+    BootEntry->BootFilePath = NULL;
+
+    BootEntry->OsOptionsLength = sizeof(NTOS_OPTIONS);
+    RtlCopyMemory(Options->Signature,
+                  NTOS_OPTIONS_SIGNATURE,
+                  RTL_FIELD_SIZE(NTOS_OPTIONS, Signature));
+
+    Options->OsLoadPath = ArcPath;
+
+    /* ReactOS */
+    // BootEntry->BootEntryKey = MAKESTRKEY(L"ReactOS");
```

Comment:
I've not yet fully decided how I want to specify what I call the "Boot Entry Key" (is actually some kind of internal name for the boot entry, e.g. in freeldr.ini it's a boot entry section name, but in case of the registry BCD it would be some GUID I suppose).